### DISCUSSION TITLE
Use HomeAssistant sensors inside of RasperryMatic / CCU3

### DISCUSSION THREAD

**Discussion by simllll** [OpenCCU/OpenCCU#2927] [🇺🇸 Q&A] [answered]
Hello everyone,
I've successfully set up the RaspberryMatic AddOn inside of Home Assistant. I also got the RFUSB stick working and paired the first devices. So far, so good.

But now I've hit the first problem. After installing the HmIP-FALMOT-C12 (floor heating actor), I discovered that I can't operate it without a temperature sensor. While I can briefly open the actuators, they immediately revert due to some internal routines (couldn't find out what was triggering it) and close again. (They work fine and I even can overwrite them as soon as I add a thermostat to the heating group). However, I have several temperature sensors in my apartment (e.g., from the air conditioning units or various Zigbee sensors), and I was thinking about the following scenario:

* Create a virtual thermostat in the CCU3 / RaspberryMatic (HM-IP).
* Set up a Home Assistant automation that updates the sensor whenever it changes.
* Use the "virtual device" inside of CCU3 like a real one and combine it with the relevant heating groups.

Specifically, I'd like to apply this for the opening sensors, while for temperature, I'll likely end up installing Homematic thermostats anyway.

TL;DR: How could I create a "fake" / virtual device in RaspberryMatic that I can feed with real data via Home Assistant automations, so that I can use my existing devices inside RasperryMatic?

I've already tried [Homematic Virtual Interface](https://github.com/thkl/Homematic-Virtual-Interface) and the [CuX-Daemon](https://github.com/jens-maus/cuxd), but without success. With the HVI, I can only use plugins, but I'm unable to create a custom/virtual device. The CuX Daemon sounded promising, but unfortunately, there are no HM-IP devices. I can only create "HM" devices, which I can't combine with my HM-IP devices, e.g., in groups.

Thanks for any input! 💯

Simon

// This is the Englisch version, German version is here: https://github.com/jens-maus/RaspberryMatic/discussions/2926

**Comment by jens-maus** [comment, marked as answer]
> But now I've hit the first problem. After installing the HmIP-FALMOT-C12 (floor heating actor), I discovered that I can't operate it without a temperature sensor. While I can briefly open the actuators, they immediately revert due to some internal routines (couldn't find out what was triggering it) and close again. (They work fine and I even can overwrite them as soon as I add a thermostat to the heating group). However, I have several temperature sensors in my apartment (e.g., from the air conditioning units or various Zigbee sensors), and I was thinking about the following scenario:
> 
> * Create a virtual thermostat in the CCU3 / RaspberryMatic (HM-IP).
> * Set up a Home Assistant automation that updates the sensor whenever it changes.
> * Use the "virtual device" inside of CCU3 like a real one and combine it with the relevant heating groups.
> 
> Specifically, I'd like to apply this for the opening sensors, while for temperature, I'll likely end up installing Homematic thermostats anyway.

Unfortunately, non of your ideas/suggestions will work. The HmIP-FALMOT and all other Homematic Floor actuators **require** a coupled wall thermostat. There is no possibility to create a virtual thermostat, automation or anything like that to manually "drive" these type of floor heating actuators. They can only operate with a coupled wall thermostat from HomematicIP because they can only perform their internal and automated operations together with their wall thermostats. Furthermore, there is no possibility to explicitly open or close one channel / motor of these floor actuators because all this is done solely via internal logic between the floor actuator and the wall thermostat. All you can do is to set a certain target temperature at the wall thermostat and then the thermostat and the floor actor will perform their internal logic until this temperature is reached.

**Reply by simllll** [reply]
Thanks for your response, the CuX-Daemon can create HM devices, but not HM-IP devices... as your are the author, you also think there is no way to actually add this kind of feature?

Regarding window openings, I found a blueprint (https://community.home-assistant.io/t/homematic-ip-local-reduce-room-temperature-when-a-window-or-door-is-opened/373305) and adapted it a bit:
<details>

<summary>
On Window/Door Openings notify rasperryamatic.
</summary>

blueprints/automation/ccu3/doorsensors.yaml

```yaml
blueprint:
  name: Window Open, Notify HomeMatic CCU
  description: >
    An automation blueprint that reduces the set temperature of your climate
    device or group if a window sensor is open for more than the preset wait time.
    It uses the Homematic window sensor channel to set the temperature to the "Open-window
    temperature" set in the device configuration. You should set the "Mode for temperature
    fall detection" to Inactive to avoid conflicts. It waits until the window is closed
    again in order to restore the climate entity temperature. It has an optional blocking
    entity to prevent the automation running unnecessarily, for example during the
    Summer season.
  domain: automation
  input:
    window_entity:
      name: Window Sensor
      description: The window or door sensor that will control the climate entity.
      selector:
        entity:
          filter:
            - domain: binary_sensor
              device_class:
                - window
                - door
          multiple: true
    open_delay:
      name: Windows Open Wait Time
      description: Time to delay after window opens before notifying HomeMatic.
      default: 30
      selector:
        number:
          min: 0.0
          max: 600.0
          unit_of_measurement: seconds
          mode: slider
          step: 1.0
    blocking_entity:
      name: (OPTIONAL) Blocking Entity
      description: >
        If this entity's state is on, it will prevent the automation from
        running. E.g. summer mode or away mode.
      default:
      selector:
        entity:
          filter:
            domain: input_boolean
    climate_entity:
      name: Climate Device
      description: The climate entity that is controlled by the window sensor.
      selector:
        device:
          entity:
            - integration: homematicip_local
  source_url: https://community.home-assistant.io/t/homematic-ip-local-reduce-room-temperature-when-a-window-or-door-is-opened/373305
variables:
  blocking_entity: !input "blocking_entity"
  climate_target: !input "climate_entity"
  climate_id: !input "climate_entity"
  open_delay: !input "open_delay"
  windows: !input "window_entity"
trigger:
  - trigger: state
    entity_id: !input "window_entity"
    to: "on"
    for: !input "open_delay"
condition:
  - condition: template
    value_template: "{{ (blocking_entity == none) or (states[blocking_entity].state == 'off') }}"
action:
  - action: homematicip_local.set_device_value
    data:
      device_id: "{{ climate_id }}"
      channel: 1
      parameter: WINDOW_STATE
      value: 1
  - wait_template: >
      {% set closed = namespace(bool=true) %}
      {% for state in states.binary_sensor%}
          {% if state.entity_id in windows %}
              {% if state.state == "on" %}
               {% set closed.bool = false %}
              {% endif %}
         {% endif %}
      {% endfor %}
      {{closed.bool}}
    continue_on_timeout: false
  - action: homematicip_local.set_device_value
    data:
      device_id: "{{ climate_id }}"
      channel: 1
      parameter: WINDOW_STATE
      value: 0
mode: restart
max_exceeded: silent

```
</details>
It's not perfect, but it works ;-)

**Reply by simllll** [reply]
One more thign regarding floor actuuactors requiring a thermostat, ... is there any possibility to programtically change the "group" settings of a heating group? In my case I have a ceiling cooling system and a floor heating system. When I switch to summer mode, I would like to control the ceiling cooling system with the same thermostats.
I didn't know about this limitations that every actuator needs a thermostat to work, that's kind of a problem now in my setup.

**Reply by jens-maus** [reply]
> Thanks for your response, the CuX-Daemon can create HM devices, but not HM-IP devices... as your are the author, you also think there is no way to actually add this kind of feature?

No, this won't work either. As said, the HmIP-FALMOT (and any other eQ3 floor actuator) requires a homematicIP wall thermostat in hardware coupled to them to work properly and with the internal logic of the floor actuator. There is no possibility to define or modify the valve opening state of these floor actuators in software, neither via HA nor via CuXD. In fact, such a homematicIP wall thermostat and the floor actuator or directly coupled so that even if your whole LAN or HA / RaspberryMatic is down, they still communicate to eachother via their internal 868MHz link (Direktverknüpfung). That's in fact the whole speciality / feature behind homematicIP, that they work without any central at all and if there is a central then you modify the target temperature of the wall thermostat via HA/RaspberryMatic and then the wall thermostat will communicate with the floor actuator to reach this target with its internal heating logic...

**Reply by simllll** [reply]
> they still communicate to eachother via their internal 868MHz link (Direktverknüpfung). That's in fact the whole speciality / feature behind homematicIP

Thanks, I saw that already while looking into the  "Routing"-Option. And reading about "Advanced Routing". That also sounded like it can talk to the "base" as fallback if the other device is not directly reachable (e.g. HM-Ip Wireless vs Wired), and there I thought it could be possible to bring in a "virtual device".

**Reply by simllll** [reply]
> One more thign regarding floor actuuactors requiring a thermostat, ... 

opened another thread for this topic over here https://github.com/danielperna84/hahomematic/discussions/1858

**Reply by MichaelN0815** [reply]
Hast Du echt angenommen ein Heizkreis Verteiler kann ohne einen Input seine Arbeit verrichten?

**Reply by simllll** [reply]
> Hast Du echt angenommen ein Heizkreis Verteiler kann ohne einen Input seine Arbeit verrichten?

Nein, aber ich hab angenommen ich kann vorhanden Inputs re-usen. So wie ungefähr überall anders ;-) Ich brauch ja nicht 5 Temperaturmessgeräte in einem Raum.

**Reply by jens-maus** [reply]
> > Hast Du echt angenommen ein Heizkreis Verteiler kann ohne einen Input seine Arbeit verrichten?
> 
> Nein, aber ich hab angenommen ich kann vorhanden Inputs re-usen. So wie ungefähr überall anders ;-) Ich brauch ja nicht 5 Temperaturmessgeräte in einem Raum.

Dann nutz nur ein HmIP-WTH als dein Hauptthermostat in dem jeweiligen Raum und entferne den den du da jetzt vermeintlich bereits hast. Anders wird es nicht gehen, egal wie oft du es probierst oder versuchst anders zu lösen. Oder ersetze den HmIP-FALMOT eben gegen etwas anderes wenn diese Limitation dir zu einschneidend ist.

**Comment by simllll** [comment]
Ui I got close, but still failed :-1: 

Looking into https://github.com/mdzio/ccu-jack I even managed to create a virtual HmIP device:
![image](https://github.com/user-attachments/assets/baffc02c-03af-4e90-ad66-bf59df430862)

The last three are virtual thermostats. I can update the values through MQTT or direclty via HA, that would be perfect..  BUT, I still not able to use this thermostat in a heating group :( .  Guess this is the limitation @jens-maus was talking about, or am I missing something else here?

**Reply by jens-maus** [reply]
> Guess this is the limitation @jens-maus was talking about, or am I missing something else here?

As I told you: There is *NO* possibility to use a software-based HmIP wall thermostat replacement (also not via CCU-Jack) to couple it to a HmIP-FALMOT or other HmIP floor actuator. The FALMOT *requires* a HmIP thermostat (HmIP-WTH) in *hardware* to work correctly. Without that thermostat this floor heating actuator is not able to get the temperature information and thus will not work. And it also does not have any other channel to get or assign it the required temperature information. So either get a HmIP wall thermostats (HmIP-WTH) and use that as your main thermostat for the desired rooms and then use that from HA or replace the HmIP-FALMOT with something completely different. No other options exist.

**Reply by simllll** [reply]
Thanks, so far I see the current limitations. As I still like to play around with it, why do you think emulating a hardware thermostat is not feasible? Virtualization is done with so many things today, I just can't figure out why it should not be doable with a thermostat based on the "ipv6" protocol (right, I couldn't find any documentation about the homematic ip protocol yet, but I guess there are ways to find that out too?)

**Reply by jens-maus** [reply]
> As I still like to play around with it, why do you think emulating a hardware thermostat is not feasible? Virtualization is done with so many things today, I just can't figure out why it should not be doable with a thermostat based on the "ipv6" protocol (right, I couldn't find any documentation about the homematic ip protocol yet, but I guess there are ways to find that out too?)

Well, the counter question would be: Why do you ask questions when you cannot accept a definite answer from someone with deeper technical knowledge about something you obviously have just started to look at? So I can only repeat myself: What you want is **NOT** possible without eQ3 changing the device firmware of the HmIP-FALMOT and allowing to simply provide the necessary temperature data in a different manner. And the IP term in "HomematicIP" does not mean anything even if you think this is a standard "ipv6" protocol. In fact, the IP in the HomematicIP naming is just a marketing term even thought they are internally using something like IPv6 within their RF communication between there devices. However, as said: This does not mean you can send any command or perform any IP based communication to these devices. So please make yourself more familiar if you feel thus must be possible to some extend. And I wouldn't be surprised if you then soon identify yourself that what you actually want (couple a thermostat from a different vendor or use a software-based thermostat) is simply not possible.

**Reply by simllll** [reply]
Hi @jens-maus , sorry if it sounded like an offense against you, this was not my intention at all. It's just that I have experienced in the last 2-3 years the whole world of home assistant and other open solutions, and wasn't expecting that homematic is limited/closed in so many ways. My idea, due to the open hardware for homematic that is available and the rasperrymatic project, was that homematic is open to customization and co too. Which is unfortunately not the case. It's more the opposite, you can do what they want you to do, but any special case is not possible at all. (Like it is in the open source and home assistant world ;-)). 
Anyways, I could solve my issues. Returned some homematic hardware and replaced it with zigbee and other open protocols. I have now a combination of homematic for my floor heating and a zigbee controller for my ceiling cooling system. With some "hacks" / additional automations I was able to bring homematic to the level of customization I needed for this setup. So in the end, fine enough for me. 

**Thanks for your help and support.** All the best for the homematic future, hopefully they become more open (source) in the end too ;-)
